keps/sig-storage/20191125-bucket-provisioning.md

```diff
@@ -0,0 +1,540 @@
++ Use native Kubernetes resources where possible, again, to keep the bucket experience for users and admins similar to existing storage provisioning.
++ Be unopinionated about the underlying object-store and at the same time provide a flexible API such that provisioner specific features can be supported.
++ Ensure bucket consuming pods wait until the target buckets have been created and are accessible. Thus there is no specific order required for when a bucket claim is created vs. when the app pod is run.
++ Present similar user and admin experiences for both _greenfield_ (new) and _brownfield_ (existing) bucket provisioning.
+
+### Non-Goals
+
++ Update the native Kubernetes PVC-PV API to support object buckets.
++ Define a native _data-plane_ object store API.
++ Handle the small percentage of apps that will not be portable due to use of non-compatible object-store features.
+
+## Proposal
+
+Create a bucket provisioning library, similar to the Kubernetes [sig-storage-lib-external-provisioner](https://github.com/kubernetes-sigs/sig-storage-lib-external-provisioner/blob/master/controller/controller.go) library, which handles all of the controller and Kubernetes resources related to provisioning.
+This library will implement a simple and familiar API via Custom Resources, and define a _contract_ between app developers and provisioners regarding the Kubernetes resources related to bucket provisioning.
+
+The actual creation of physical buckets and the generation of appropriate credentials belong to each object store provisioner, while the library handles watches on bucket claims, reconciles desired state, and creates and deletes k8s resources.
+
+Since certain object store-specific artifacts, e.g. user name, are needed in order to clean up after a bucket is deleted or revoked, this proposal must allow for some provisioner values to be stored across API calls.
+
+## Design Details
+
+### Custom Resources
+
+(see the [full specs](#custom-resource-details))
+
+The bucket provisioning library utilizes two Custom Resources to abstract an object store bucket and a claim/request for such a bucket.
++ a namespaced _**ObjectBucketClaim**_ (OBC), similar to a PVC, is a formal request for a new or an existing bucket.
++ an _**ObjectBucket**_ (OB), similar to a PV, is the k8s representation of the provisioned bucket and may contain object store-specific data needed to allow the bucket to be de-provisioned. The OB resides in the same namespace as the object store, and is typically not seen by bucket consumers.
+**Note:** namespaced OBs is a change from the original design and was suggested at the 2019 NA Kubecon storage face-to-face with the premise being that if there is no _technical_ reason for a cluster scoped resource then it should be namespaced.
+
+As is true for PVCs-PVs, there is a 1:1 relationship between an OBC and an OB, and as will be seen below, there is a [_binding_](#binding) between an OBC and OB. 
```

Comment:
What is the reason to have two separate objects, OBC and OB? For PV/PVC it's security. If volume ID was in PVC, user could craft their own PVCs and guess volume ID of some other volume and steal its data. In this proposal, you give full bucket ID and secrets to access it to a pod. So user can guess ID of some other bucket and steal it in the pod.

Therefore you don't need OB object at all. All can be in OBC (and it could be called ObjectBucket instead of claim).

Reply:
A reason for the OB is to contain provisioner-specific data, eg. the newly created user name, so that provisioners can delete buckets and clean up policies etc.

Reply:
Are these provisioner-specific data secret? In other words, can a malicious user misuse it to revoke / delete data of someone else?

Reply:
@jsafrane The secret is created by the lib and lives in the same namespace as the obc. The secret contains s3-like access and secret key values, and a provider-specific _map[string]string_.
The secret cannot be used maliciously AFAICT, because the provisioner is expected to return creds that are specific to a single bucket and are not the owner's creds.

Reply:
@jsafrane This is dependent on the provisioner implementation.  The library only passes data returned from Provision()/Grant() calls directly into the secret.  Rook, for instance, creates a new Ceph user with limited rights and returns those credentials.  That user may only access the generated/attached bucket, may not create buckets, and (as a design of Ceph), may not view or mutate other user's buckets.